src/NumericalAlgorithms/Interpolation/RegularGridInterpolant.hpp

```diff
@@ -0,0 +1,240 @@
+
+#pragma once
+
+#include <algorithm>
+#include <array>
+#include <cstddef>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Matrix.hpp"
+#include "DataStructures/Variables.hpp"
+#include "ErrorHandling/Assert.hpp"
+#include "NumericalAlgorithms/LinearOperators/Transpose.hpp"
+#include "Utilities/Blas.hpp"
+#include "Utilities/Gsl.hpp"
+
+// IWYU pragma: no_forward_declare Variables
+
+/// \cond
+template <size_t Dim>
+class Mesh;
+namespace PUP {
+class er;
+}  // namespace PUP
+/// \endcond
+
+namespace intrp {
+
+/// \ingroup NumericalAlgorithmsGroup
+/// \brief Interpolate a Variables from a Mesh onto a regular grid of points.
+///
+/// The target points must lie on tensor-product grid that is aligned with the
+/// source Mesh. In each dimension of the target grid, however, the points can
+/// be freely distributed; in particular, the grid points need not be the
+/// collocation points corresponding to a particular basis and quadrature.
+///
+/// \note When interpolating between two different resolutions on the same type
+/// of spectral grid (e.g., between two Legendre Gauss Lobatto grids with
+/// different numbers of points), it may be more efficient to use a projection
+/// instead of the interpolation that is done by this class. This class is
+/// intended to handle interpolation onto more general (but still regular)
+/// target grids.
+///
+/// \note The target grid need not overlap the source grid. In this case,
+/// polynomial extrapolation is performed, with order set by the order of the
+/// basis in the source grid. The extapolation will be correct but may suffer
+/// from reduced accuracy, especially for higher-order polynomials.
+template <size_t Dim>
+class RegularGrid {
+ public:
+  /// \brief An interpolator between two overlapping meshes.
+  RegularGrid(const Mesh<Dim>& source_mesh,
+              const Mesh<Dim>& target_mesh) noexcept;
+  /// \brief An interpolator from a Mesh to a regular grid of points.
+  RegularGrid(
+      const Mesh<Dim>& source_mesh,
+      const std::array<DataVector, Dim>& target_1d_logical_coords) noexcept;
+  RegularGrid();
+
+  // clang-tidy: no runtime references
+  void pup(PUP::er& p) noexcept;  // NOLINT
+
+  /// \brief Interpolate Variables onto new mesh.
+  //@{
+  template <typename TagsList>
+  void interpolate(gsl::not_null<Variables<TagsList>*> result,
+                   const Variables<TagsList>& vars) const noexcept;
+  template <typename TagsList>
+  Variables<TagsList> interpolate(const Variables<TagsList>& vars) const
+      noexcept;
+  //@}
+
+ private:
+  template <size_t LocalDim>
+  friend bool operator==(const RegularGrid<LocalDim>& lhs,
+                         const RegularGrid<LocalDim>& rhs) noexcept;
+
+  std::array<Matrix, Dim> interpolation_matrices_;
+};
+
+template <>
+template <typename TagsList>
+void RegularGrid<1>::interpolate(
+    const gsl::not_null<Variables<TagsList>*> result,
+    const Variables<TagsList>& vars) const noexcept {
```

Comment:
Most of this file appears to be a reimplementation of `apply_matrices`.

Reply:
It does appear that way :-/